packages/vats/src/lib-board.js

```diff
@@ -60,81 +177,144 @@ function makeBoard(
+      has(val) {
+        const { state } = this;
+        return state.valToId.has(val);
+      },
+      ids() {
+        const { state } = this;
+        // XXX O(n) compute for O(1) size message XXX
+        return harden([...state.idToVal.keys()]);
+      },
+      /**
+       * Get a marshaller that implements the convertValToSlot hook using getId(),
+       * "publishing" previously un-mapped objects.
+       */
+      getPublishingMarshaller() {
+        return this.facets.publishingMarshaller;
+      },
+      /**
+       * Get a marshaller that implements the convertValToSlot hook using getId(),
+       * only for objects that the board has().
+       *
+       * For other objects, we don't throw, but we emit a null slot.
+       * Un-serializing a null slot always produces a fresh object
+       * rather than preserving object identity.
+       */
+      getReadonlyMarshaller() {
+        return this.facets.readonlyMarshaller;
+      },
     },
-    getValue: id => {
-      assert.typeof(id, 'string', X`id must be string: ${id}`);
-      id.startsWith(prefix) || Fail`id must start with ${prefix}: ${id}`;
-      const digits = id.slice(prefix.length);
-      digits.match(DIGITS_REGEXP) ||
-        Fail`id must end in at least ${q(crcDigits + 1)} digits: ${id}`;
-      const seq = digits.slice(crcDigits);
-      const allegedCrc = digits.slice(0, crcDigits);
-      const crcInput = `${prefix}${seq}`;
-      const crc = calcCrc(crcInput, crcDigits);
-      allegedCrc === crc ||
-        Fail`id is probably a typo, cannot verify CRC: ${id}`;
-      idToVal.has(id) || Fail`board does not have id: ${id}`;
-      return idToVal.get(id);
+    /**
+     * @returns {ReturnType<import('@endo/marshal').makeMarshal>}
+     */
+    readonlyMarshaller: {
+      serialize(val) {
+        return my(this).readonlyMarshaller.serialize(val);
+      },
+      unserialize(data) {
+        return my(this).readonlyMarshaller.unserialize(data);
+      },
     },
-    // Adapter for lookup() protocol.
-    lookup: async (...path) => {
-      if (path.length === 0) {
-        return board;
-      }
-      const [first, ...rest] = path;
-      const firstValue = board.getValue(first);
-      if (rest.length === 0) {
-        return firstValue;
-      }
-      return E(firstValue).lookup(...rest);
+    publishingMarshaller: {
+      /** @param {*} val */
+      serialize(val) {
+        return my(this).publishingMarshaller.serialize(val);
+      },
+      /** @param {{ body: string, slots: any[] }} data */
+      unserialize(data) {
+        return my(this).publishingMarshaller.unserialize(data);
+      },
     },
-    has: valToId.has,
-    ids: () => harden([...idToVal.keys()]),
   });
+};
 
-  return board;
-}
-
-export { makeBoard };
+/**
+ * @param {bigint | number} [initSequence]
+ * @param {object} [options]
+ * @param {string} [options.prefix]
+ * @param {number} [options.crcDigits]
+ */
+export const makeBoard = (initSequence = 0, options = {}) => {
```

Comment:
why keep this `makeBoard` that invents its own baggage?

if it's needed for backwards compatibility with non-upgradable we could keep it but it should have a name to distinguish from the one prepared from baggage. E.g. `makeInitialBoard`. Or if that's too much churn, tag this with `@deprecated use the maker from prepareBoard`.

Reply:
I initially tagged it `@deprecated` but after finding it used in other packages I wondered if that's really called for. It works just fine, it has the same exterior API, and deprecating it is not necessary to address the issue I'm working on. Do we really think the test consumers should be changed to supply baggage? (we would not do it in this PR in any case...)

OTOH... I don't suppose a non-upgradeable board is a product feature that we are committed to support. I dunno.

Reply:
yeah, it's that OTOH I have in mind. if we do want to support it, I think the name should make clear to the caller what they're getting. e.g. `makeNonUpgradableBoard`.

Reply:
will try `makeInMemoryBoard`

Reply:
that's not in here. Mind if I make that change before this lands?

Reply:
feel free